Add support for Hotwire Turbo Drive

ceritium opened the pull request:
It is related to this issue https://github.com/MiniProfiler/rack-mini-profiler/issues/482

I moved to the gem the @mildred's solution.

I tested it on a small demo app, as I show in the following video: 

https://user-images.githubusercontent.com/16633/121650575-fa1ddc00-ca99-11eb-95eb-fc1b4fb0698d.mov

Any recommendation for better QA?

Dainii commented:
Tested with rails (6.1.3.1) and turbo-rails (0.5.9), it seems to work quite nicely.

kbrock commented on lib/html/includes.js:
```diff
@@ -640,6 +640,17 @@ var _MiniProfiler = (function() {
   };
 
   var bindDocumentEvents = function bindDocumentEvents() {
+    document.addEventListener("turbo:before-visit", e => {
```
it looks like the other methods do not use anonymous blocks, they use functions defined. i.e.: onClickEvents, unbindDocumentEvents

Also, it looks like there is a check that `Turbolinks` is even relevant (by checking the constant defined in a gem. is there something similar you can do here?

Probably want to move this to the bottom of this method, right after Turbolinks section. That way the meat of this bindDocumentEvents method reads the same

ceritium replied:
Thanks, @kbrock. 

I refactored the callbacks as functions and moved the code under Turbolinks logic.

`Turbo` doesn't implement `supported` like `Turbolinks`, so I am just checking that `Turbo` is available.

Doing the Turbo constant check avoid adding unnecessary event listeners, but it requires that the application exposes `Turbo` globally:

```
import { Turbo } from "@hotwired/turbo-rails"
window.Turbo = Turbo
```

What do you think? What is the best approach? I guess we should mention it in the documentation.

kbrock commented on lib/html/includes.js:
```diff
@@ -640,6 +640,17 @@ var _MiniProfiler = (function() {
+    })
+
+    document.addEventListener("turbo:load", e => {
+      if(window.MiniProfilerContainer) {
```
could you also enhance `unbindDocumentEvents` to remove these events?
I think that will remove a memory leak.

ceritium replied:
Done.

kbrock reviewed (commented):
This looks great. but please remember:

1. I am not a rack-mini-profiler developer. I do not have merge rights. (feel free to ignore my suggestions)
2. I am not a ui developer and just looking at the code and suggesting you copy stuff.

SamSaffron commented:
This looks right to me @OsamaSayegh should we merge?

OsamaSayegh reviewed (commented):
Thank you for this PR @ceritium ❤️ I'd love to merge this, but I have a couple of comments:

Re:

> Doing the Turbo constant check avoid adding unnecessary event listeners, but it requires that the application exposes Turbo globally:
>    ```
>    import { Turbo } from "@hotwired/turbo-rails"
>    window.Turbo = Turbo
>    ```
> What do you think? What is the best approach? I guess we should mention it in the documentation.

Requiring people to do something like that feels too much magical to me 😅 Instead, I'd prefer if we added a new `enable_hotwire_turbo_support` config option (false by default) to [`lib/mini_profiler/config.rb`](https://github.com/MiniProfiler/rack-mini-profiler/blob/a93060ef9d49c5e4007d00eb8bb5a783263bb220/lib/mini_profiler/config.rb) that people who use Hotwire in their application can turn on. When it's turned on, we can use that to signal to MiniProfiler to add Hotwrite's specific bits.

For example, in `lib/html/profile_handler.js` I'd replace the `data-turbo-permanent="true"` attribute that you have now with a new placeholder `{hotwriteTurboSupport}`. The `get_profile_script` method here:

https://github.com/MiniProfiler/rack-mini-profiler/blob/a93060ef9d49c5e4007d00eb8bb5a783263bb220/lib/mini_profiler/profiler.rb#L720-L774

is what replaces the placeholders in `profile_handler.js` with the right values. So, I'd add a check for our new config, and if it's true I'd set `hotwriteTurboSupport` in the `settings` hash to `data-turbo-permanent="true"`; otherwise set `hotwriteTurboSupport` to the empty string.

Finally, instead of checking for the `Turbo` constant on the client side, we can check for the `data-turbo-permanent` attribute on MiniProfiler's `<script>` tag and add our event listeners if the attribute exists.

The advantage of this is it allows to conditionally add the `data-turbo-permanent="true"` attribute only for the people who use Turbo, and avoid magical hacks like `window.Turbo = Turbo`.

One last thing, could run the `update_asset_version` rake task when you're done? This updates the `ASSET_VERSION` constant which use to bust cache when new versions are released.

Let me know if you have any questions!

ceritium commented:
Thanks for your feedback @OsamaSayegh. I like your approach. I just pushed a commit addressing your feedback and a small test. 

Would you add more tests?
Do you have any suggestions about variable naming? 
It's ok the place for `enable_hotwire_turbo_drive_support` on "config.rb" or would you move it to a different place?

OsamaSayegh commented:
I think it looks great, thank you @ceritium!

ceritium commented:
@OsamaSayegh I added the documentation. 

Do you like the commits as they are? Would you squash them? I didn't found any policy about that.

OsamaSayegh commented:
Perfect, thank you for all your work on this!

Github has a merge option that squashes everything together which I'm going to use now!
